Feat/basic app test

Fenriir42 opened the pull request:
## Description

Added core audio playback functionality, song library management system, and a dual-theme UI for the karaoke app.

## Changes

### Audio System
- File loading with multi-format support (MP3, FLAC, WAV, OGG, M4A, AAC)
- Real-time playback position tracking
- Progress bar with MM:SS duration display
- Fixed playback initialization bug

### Library Management
- Recursive directory scanner for audio files
- Search/filter to find tracks quickly
- Click-to-play from library list
- Track numbering and format badges

### UI/UX
- Theme system with "Iron Flower" (dark) and "White Frame" (light)
- Theme switcher in top panel
- Technical/industrial styling with color-coded controls
- Animated progress bar in footer

## Testing

- Tested file loading with various formats
- Verified playback timing accuracy
- Tested library scanning with nested directories
- Confirmed theme switching works correctly
- Ran `cargo fmt` and `cargo clippy` - all checks pass

## Technical Notes

- Duration extraction uses symphonia probe for accurate timing
- Playback state management handles pause/resume correctly
- error handling in audio sink initialization
- Applied clippy suggestions for cleaner code

## Breaking Changes

None - all changes are additive.

## Screenshots
<img width="1920" height="1032" alt="{4B070277-5612-45CF-A3A0-37617B749E4E}" src="https://github.com/user-attachments/assets/625a1237-714d-4ae0-a33c-cc1f49c633d2" />

-----

<img width="1920" height="1032" alt="{83A315E9-DCF2-4CAF-99D9-335B702B16C1}" src="https://github.com/user-attachments/assets/c536a5a4-9953-4b1e-b033-3b9853050226" />

Fenriir42 committed:
feat: make PyO3/Spleeter optional for cross-compilation

- Add [features] section with spleeter feature
- Make pyo3 dependency optional
- Default build excludes Python integration
- Enables Windows cross-compilation without Python complexity

Fenriir42 committed:
build: add Windows cross-compilation configuration

- Configure MinGW linker for x86_64-pc-windows-gnu target
- Enable building Windows .exe from Linux/WSL
- Required mingw-w64 toolchain and rustup target

Fenriir42 committed:
docs: update dependencies and add build tutorials

- Document optional PyO3/Spleeter feature
- Add comprehensive build instructions for Linux and Windows
- Include cross-compilation prerequisites and commands
- Add build summary table with all variants
- Update dependency versions and descriptions
- Clarify when Python is required vs optional

Fenriir42 committed:
feat: implement basic GUI application with eframe

- Initialize tracing for structured logging
- Configure eframe window with 800x600 viewport
- Set up application entry point with KaraokeApp
- Clean up build.rs (remove unused std::env import)

Fenriir42 committed:
feat: add main application state and UI update loop

- Implement KaraokeApp struct with counter, text input, audio player
- Add graceful audio initialization with fallback
- Integrate UI panels and widgets for modular rendering
- Handle audio playback with beep generation
- Track playing state and update UI accordingly

Fenriir42 committed:
feat: implement audio playback and generation module

- Add AudioPlayer with graceful fallback for missing devices
- Implement OutputStream and Sink management with rodio
- Create sine wave beep generator with WAV encoding
- Generate 440Hz A4 test tone for audio verification
- Handle audio errors with tracing logs

Fenriir42 committed:
feat: implement modular UI components with panels and widgets

- Add top panel with application title and branding
- Add bottom panel with playback status and counter display
- Create reusable widgets: text input, counter, audio controls
- Implement AudioAction enum for clean audio event handling
- Add informational section with demo instructions
- Use colored labels for visual status indicators

Fenriir42 committed:
feat: add audio file loader with symphonia

- Create loader module for loading MP3, FLAC, WAV, OGG, M4A, AAC files
- Use anyhow for error handling with context
- Return rodio Decoder ready for playback
- Add format validation and error formatting helpers
- Supports all symphonia-compatible formats

Fenriir42 committed:
feat: integrate file picker and audio file playback

- Add rfd file dialog for selecting audio files
- Track current file and error messages in app state
- Implement load_and_play_file with error handling
- Add toggle_playback for pause/resume functionality
- Clear previous audio before loading new file
- Display errors in UI when file loading fails

Fenriir42 committed:
feat: add file playback UI controls

- Create render_file_playback_section with Open/Play/Pause/Stop
- Display currently playing filename
- Show error messages in red when file loading fails
- Add OpenFile and PlayPause actions to AudioAction enum
- Disable play button when no file is loaded
- Show helpful hint when no file is selected

Fenriir42 committed:
feat(library): add song library with directory scanning

- Add Song struct with path, name, and extension fields
- Implement directory scanner using walkdir
- Support MP3, WAV, FLAC, OGG, M4A, and AAC formats
- Sort songs alphabetically by name

Fenriir42 committed:
feat(audio): add playback position tracking and duration display

- Add timing fields to AudioPlayer (start_time, pause_time, accumulated_time)
- Implement start_tracking(), get_position(), and reset_position() methods
- Update pause/resume/stop/clear to manage playback timing
- Add get_audio_duration() to extract duration using symphonia probe
- Fix playback start issue by adding explicit sink.play() call

Fenriir42 committed:
feat(ui): add theme system with Iron Flower and White Frame palettes

- Create Theme enum with IronFlower (dark) and WhiteFrame (light) modes
- Implement color methods: primary(), active(), text_primary(), text_muted()
- Add theme.apply() to update egui visuals with custom backgrounds
- Add theme switcher button in top panel
- Theme colors follow design guide specifications

Fenriir42 committed:
feat(ui): redesign with IBO-inspired technical styling

- Update top panel with technical header and system indicators
- Add real-time progress bar with MM:SS duration display
- Redesign playback controls with military-style buttons (⊡, ▶, ⊟, ⊠)
- Add library database section with track numbers and filter
- Apply theme colors throughout all UI components
- Remove hardcoded colors in favor of theme system

Fenriir42 committed:
feat(app): integrate library and theme system

- Add library fields (Vec<Song>, library_path, library_filter)
- Add theme state with IronFlower as default
- Implement scan_library() with folder picker
- Update load_and_play_file() to extract duration and start tracking
- Handle theme switching with logging
- Pass theme and timing data to UI components
- Wire up library actions (scan, play song)

Ciznia committed:
chore: remove cargo lock from gitignore

Ciznia commented on src/audio/player.rs:
```diff
@@ -0,0 +1,137 @@
+            },
+        };
+
+        let sink = match Sink::try_new(&stream_handle) {
```
what is a Sink and where does it come from

Ciznia commented on src/library/mod.rs:
```diff
@@ -0,0 +1,36 @@
+}
+
+impl Song {
+    /// Create a new Song from a path
```
Make useful documentation.
What are the parameter type, and which error handling is done.
For exemple here if path point to an non existing file, what does the function do. It's not clear in code neither in documentation

Ciznia commented on src/library/scanner.rs:
```diff
@@ -0,0 +1,62 @@
+        let entry_path = entry.path();
+
+        // Skip directories
+        if !entry_path.is_file() {
+            continue;
+        }
```
Maybe scan recursively to allow subfolder in the library folder

Fenriir42 replied:
will be done with the overhaul of the UI

Ciznia commented on src/main.rs:
```diff
@@ -1,8 +1,27 @@
-// PWE Karaoke - Main entry point
-// This is just a placeholder to ensure the project structure is set up correctly
+mod app;
+mod audio;
+mod library;
+mod ui;
```
Maybe it's my oversight from the rustling exercices but does we need to define all the mod here even the one we don't use in this file

Fenriir42 replied:
nope we can't because the modules are at the crate root level (`src`), not submodules of `app`
```sh
    Checking pwe-karaoke v0.1.0 (/home/julien/delivery/tek3/PRO/pwe)
error[E0583]: file not found for module `audio`
 --> src/app.rs:1:1
  |
1 | mod audio;
  | ^^^^^^^^^^
  |
  = help: to create the module `audio`, create file "src/app/audio.rs" or "src/app/audio/mod.rs"
  = note: if there is a `mod audio` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `library`
 --> src/app.rs:2:1
  |
2 | mod library;
  | ^^^^^^^^^^^^
  |
  = help: to create the module `library`, create file "src/app/library.rs" or "src/app/library/mod.rs"
  = note: if there is a `mod library` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `ui`
 --> src/app.rs:3:1
  |
3 | mod ui;
  | ^^^^^^^
  |
  = help: to create the module `ui`, create file "src/app/ui.rs" or "src/app/ui/mod.rs"
  = note: if there is a `mod ui` elsewhere in the crate already, import it with `use crate::...` instead

error[E0433]: failed to resolve: could not find `audio` in the crate root
 --> src/app.rs:9:12
  |
9 | use crate::audio::{generator, loader, player::AudioPlayer};
  |            ^^^^^ could not find `audio` in the crate root

error[E0432]: unresolved import `crate::audio`
 --> src/app.rs:9:12
  |
9 | use crate::audio::{generator, loader, player::AudioPlayer};
  |            ^^^^^ could not find `audio` in the crate root

error[E0432]: unresolved import `crate::library`
  --> src/app.rs:10:12
   |
10 | use crate::library::{scanner, Song};
   |            ^^^^^^^ could not find `library` in the crate root

error[E0433]: failed to resolve: could not find `ui` in the crate root
  --> src/app.rs:11:12
   |
11 | use crate::ui::{panels, theme::Theme, widgets};
   |            ^^ could not find `ui` in the crate root

error[E0432]: unresolved import `crate::ui`
  --> src/app.rs:11:12
   |
11 | use crate::ui::{panels, theme::Theme, widgets};
   |            ^^ could not find `ui` in the crate root

error: file is loaded as a module multiple times: `/home/julien/delivery/tek3/PRO/pwe/.clippy.toml`
 --> src/app.rs:1:1
  |
1 | mod audio;
  | ^^^^^^^^^^ first loaded here
2 | mod library;
  | ^^^^^^^^^^^^ loaded again here
3 | mod ui;
  | ^^^^^^^ loaded again here
  |
  = help: replace all but one `mod` item with `use` items
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#duplicate_mod
  = note: `-D clippy::duplicate-mod` implied by `-D clippy::suspicious`
  = help: to override `-D clippy::suspicious` add `#[allow(clippy::duplicate_mod)]`

Some errors have detailed explanations: E0432, E0433, E0583.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `pwe-karaoke` (bin "pwe-karaoke") due to 9 previous errors
```

Ciznia commented on src/ui/panels.rs:
```diff
@@ -0,0 +1,122 @@
+            ui.colored_label(theme.primary(), "●");
+            ui.add_space(5.0);
+
+            // Main title with technical styling
```
Useless comment, the call is clear enough. Even with no knowledge on the library, we can guess what these calls do

Ciznia commented on src/ui/theme.rs:
```diff
@@ -0,0 +1,80 @@
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Theme {
+    IronFlower, // Dark mode
+    WhiteFrame, // Light mode
```
Why these name and not just light/dark, and since it's a public thing, we maybe want to document it

Fenriir42 replied:
explained in the design guide, just not pushed on this branch

Ciznia commented on src/ui/theme.rs:
```diff
@@ -0,0 +1,80 @@
+}
+
+impl Theme {
+    pub fn toggle(self) -> Self {
+        match self {
+            Theme::IronFlower => Theme::WhiteFrame,
+            Theme::WhiteFrame => Theme::IronFlower,
+        }
+    }
```
Maybe use [that crate](https://github.com/Zij-IT/EnumCycling) to do a proper cycle between the enum option that way it would be simpler to add another enum without having to worry about the toggle fonction needing a rewrite

Fenriir42 replied:
let's keep this consideration for the UI refactor

Ciznia commented on src/ui/widgets.rs:
```diff
@@ -0,0 +1,345 @@
+            ui.colored_label(theme.active(), format!("[{} TRACKS]", library.len()));
+        } else {
+            ui.label(
+                egui::RichText::new("[NO DATABASE LOADED]")
```
That string would be clearer about the error
```suggestion
                egui::RichText::new("Library folder doesn't exist")
```

Ciznia commented on src/ui/widgets.rs:
```diff
@@ -0,0 +1,345 @@
+            .show(ui, |ui| {
+                if filtered_songs.is_empty() {
+                    ui.label(
+                        egui::RichText::new("[NO MATCHES FOUND]")
+                            .color(theme.text_muted())
+                            .italics(),
+                    );
+                } else {
+                    for (idx, song) in filtered_songs.iter().enumerate() {
+                        ui.horizontal(|ui| {
+                            // Track number
+                            ui.label(
+                                egui::RichText::new(format!("{:03}", idx + 1))
+                                    .color(theme.text_muted())
+                                    .small(),
+                            );
+
+                            ui.add_space(5.0);
+
+                            // Song name button
+                            let track_btn = egui::Button::new(
+                                egui::RichText::new(&song.name).color(theme.text_primary()),
+                            );
+                            if ui.add(track_btn).clicked() {
+                                action = LibraryAction::PlaySong(song.path.clone());
+                            }
+
+                            // File format badge
+                            ui.label(
+                                egui::RichText::new(format!("[{}]", song.extension.to_uppercase()))
+                                    .small()
+                                    .color(theme.text_muted()),
+                            );
+                        });
+                    }
+                }
```
Use an early return to avoid tabing of that
```suggestion
                if filtered_songs.is_empty() {
                    ui.label(
                        egui::RichText::new("[NO MATCHES FOUND]")
                            .color(theme.text_muted())
                            .italics(),
                    );
                    return action
                }
                  for (idx, song) in filtered_songs.iter().enumerate() {
                      ui.horizontal(|ui| {
                          // Track number
                          ui.label(
                              egui::RichText::new(format!("{:03}", idx + 1))
                                  .color(theme.text_muted())
                                  .small(),
                          );

                          ui.add_space(5.0);

                          // Song name button
                          let track_btn = egui::Button::new(
                              egui::RichText::new(&song.name).color(theme.text_primary()),
                          );
                          if ui.add(track_btn).clicked() {
                              action = LibraryAction::PlaySong(song.path.clone());
                          }

                          // File format badge
                          ui.label(
                              egui::RichText::new(format!("[{}]", song.extension.to_uppercase()))
                                  .small()
                                  .color(theme.text_muted()),
                          );
                      });
                  }
```

Ciznia reviewed (changes requested):
Also it would be better for readability, it your definition need is from up to down for exemple moving `LibaryAction` and `AudioAction` to the top of the file since it doesn't need anything and is needed by `render_<name>_section` and this exemple hold true for a lot of your code

Fenriir42 committed:
refactor: reorganize type definitions for better readability

Move AudioAction and LibraryAction enums to top of widgets.rs
Type definitions now appear before functions that use them
Improves code readability by following top-to-bottom dependency order

Fenriir42 committed:
refactor: enhance documentation and clean up comments across audio and UI modules

Fenriir42 committed:
fix(ui): update library section to handle missing folder case
